compiler/semstmts.nim

```diff
@@ -2178,6 +2178,13 @@ proc semProcAux(c: PContext, n: PNode, kind: TSymKind,
 
   if sfBorrow in s.flags and c.config.cmd notin cmdDocLike:
     result[bodyPos] = c.graph.emptyNode
+  
+  if sfVirtual in s.flags:
+    if c.config.backend == backendCpp:
+      let typ = s.typ.sons[1].sons[0] #TODO error if not a ptr or should we allow ref/objects?
```

Comment:
You need to attach the virtual to a **nominal** type! The compiler is free to produce as many `ptr X` types as it pleases because these are structural. You cannot attach methods to structural types and expect a consistent outcome.